File: backend/src/test/java/com/ll/hotel/domain/member/member/service/BusinessServiceTest.java
package com.ll.hotel.domain.member.member.service;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.ll.hotel.domain.member.member.dto.request.BusinessRequest;
import com.ll.hotel.domain.member.member.entity.Business;
import com.ll.hotel.domain.member.member.entity.Member;
import com.ll.hotel.domain.member.member.entity.Role;
import com.ll.hotel.domain.member.member.repository.MemberRepository;
import com.ll.hotel.domain.member.member.type.BusinessApprovalStatus;
import com.ll.hotel.domain.member.member.type.MemberStatus;
import java.time.LocalDate;
import java.util.ArrayList;

import com.ll.hotel.global.exceptions.ServiceException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpStatus;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
@ActiveProfiles("test")
@Transactional
public class BusinessServiceTest {
    @Autowired
    private BusinessService businessService;

    @Autowired
    private MemberRepository memberRepository;

    private Long testId;

    @BeforeEach
    void setUp() {
        memberRepository.deleteAll();

        for (int i = 0; i < 3; i++) {
            Member member = memberRepository.save(Member
                    .builder()
                    .birthDate(LocalDate.now())
                    .memberEmail(String.format("member[%02d]", i))
                    .memberName("member")
                    .memberPhoneNumber("01012345678")
                    .memberStatus(MemberStatus.ACTIVE)
                    .role(Role.BUSINESS)
                    .oauths(new ArrayList<>())
                    .build()
            );

            if (i == 0) {
                testId = member.getId();
            }
        }
    }

    @Test
    @DisplayName("사업자 등록 - 01")
    public void registerBusinessTest() {
        // Given
        BusinessRequest.RegistrationInfo businessRequest = new BusinessRequest.RegistrationInfo(
                "1234567890",
                LocalDate.now(),
                "김사장"
        );

        Member member = memberRepository.findById(testId).get();

        // When
        Business result = businessService.register(businessRequest, member, "01");

        // Then
        assertThat(result).isNotNull();
        assertThat(result.getApprovalStatus()).isEqualTo(BusinessApprovalStatus.APPROVED);
        assertThat(result.getBusinessRegistrationNumber()).isEqualTo(businessRequest.businessRegistrationNumber());
    }

    @Test
    @DisplayName("사업자 등록 실패 - 02")
    public void registerBusinessFailureTest() {
        // Given
        BusinessRequest.RegistrationInfo businessRequest = new BusinessRequest.RegistrationInfo(
                "1234567890",
                LocalDate.now(),
                "김사장"
        );

        Member member = memberRepository.findById(testId).get();

        // When
        ServiceException exception = assertThrows(ServiceException.class, () ->
                businessService.register(businessRequest, member, "02")
        );

        // Then
        assertThat(exception.getResultCode()).isEqualTo(HttpStatus.BAD_REQUEST);
    }
}

File: backend/src/test/java/com/ll/hotel/domain/member/member/service/BusinessServiceTest.kt
package com.ll.hotel.domain.member.member.service

import com.ll.hotel.domain.member.member.dto.request.BusinessRequest
import com.ll.hotel.domain.member.member.entity.Member
import com.ll.hotel.domain.member.member.entity.Role
import com.ll.hotel.domain.member.member.repository.MemberRepository
import com.ll.hotel.domain.member.member.type.BusinessApprovalStatus
import com.ll.hotel.domain.member.member.type.MemberStatus
import com.ll.hotel.global.exceptions.ErrorCode
import com.ll.hotel.global.exceptions.ServiceException
import io.mockk.*
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate

@SpringBootTest
@ActiveProfiles("test")
@Transactional
class BusinessServiceTest (
    private val businessService: BusinessService,
    private val memberRepository: MemberRepository,
) {
    private val businessValidationService = mockk<BusinessValidationService>()
    private var testId: Long = 0L

    @BeforeEach
    fun setUp() {
        memberRepository.deleteAll()

        testId = memberRepository.save(
            Member(
                birthDate = LocalDate.now(),
                memberEmail = "member@email.com",
                memberName = "member",
                memberPhoneNumber = "01012345678",
                memberStatus = MemberStatus.ACTIVE,
                role = Role.BUSINESS
            )
        ).id
    }

    private fun createBusinessRequest() = BusinessRequest.RegistrationInfo(
        businessRegistrationNumber = "1234567890",
        startDate = LocalDate.now(),
        ownerName = "홍길동"
    )

    @Test
    @DisplayName("사업자 등록 - 01")
    fun `should register business successfully`() {
        // Given
        val businessRequest = createBusinessRequest()

        val member = memberRepository.findById(testId).get()
        every { businessValidationService.validateBusiness(any()) } just runs

        // When
        val result = businessService.register(businessRequest, member)

        // Then
        assertThat(result.approvalStatus).isEqualTo(BusinessApprovalStatus.APPROVED)
        assertThat(result.businessRegistrationNumber).isEqualTo(businessRequest.businessRegistrationNumber)

        verify { businessValidationService.validateBusiness(businessRequest) }
    }

    @Test
    @DisplayName("사업자 등록 실패 - 02")
    fun `should throw exception when business registration fails`() {
        // Given
        val businessRequest = createBusinessRequest()
        val member = memberRepository.findById(testId).get()

        every { businessValidationService.validateBusiness(any()) } answers {
            ErrorCode.INVALID_BUSINESS_INFO.throwServiceException()
        }

        // When
        val exception = assertThrows<ServiceException> {
            businessService.register(businessRequest, member)
        }

        // Then
        assertThat(exception.resultCode).isEqualTo(ErrorCode.INVALID_BUSINESS_INFO)

        verify { businessValidationService.validateBusiness(businessRequest) }
    }
}